### ISSUE TITLE
Dataset.to_batches accumulates memory usage and leaks

### ISSUE THREAD

**Issue by akoumjian** [apache/arrow#39808]
### Describe the bug, including details regarding any error messages, version, and platform.

If you want to read in a large parquet file or series of parquet files, the dataset reader accumulates the memory it allocates as you iterate through the batches.

To recreate:

```python
import pyarrow as pa
import pyarrow.parquet as pq
import pyarrow.dataset as ds

# Create dataset from large parquet file or files
dataset = ds.dataset('data/parquet', format='parquet')

# Iterate through batches
for batch in dataset.to_batches(batch_size=1000, batch_readahead=0, fragment_readahead=0):
    print(pa.total_allocated_bytes())
    pass
print(pa.total_allocated_bytes())
````

I am running this on OSX, which I believe uses `mimalloc` backend by default. It's worth noting that this is not the behavior that `ParquetFile.iter_batches` has. If you swap in that iterator, it will de-allocate the memory as soon as the batch leaves scope.

### Component(s)

Parquet, Python

**Comment by jorisvandenbossche** [comment]
Can you print the output of what you see locally? (and ideally share a full reproducer that includes creating a dummy parquet file that reproduces the issue for you, in case it is dependent on the content/characteristics of the parquet file)  
I am trying to reproduce this on my linux laptop, but I don't see a continuous increase in allocated bytes.

**Comment by daverigby** [comment]
I am observing the same thing with a single parquet file of 10M records (https://storage.googleapis.com/pinecone-datasets-dev/yfcc-10M-filter-euclidean-formatted/passages/part-0.parquet - 2.3GB).

Using code equivalent to OPs, I see `total_allocated_bytes` increase consistently over the run, requiring over 5.9GB to iterate the file.

Using `ParquetFile.iter_batches` as suggested, memory usage is much more stable (although increases a little over the duration).

Variation of the OPs reproducer code which includes the two modes:
```Python
#!/usr/bin/env python3

import pyarrow as pa
import pyarrow.parquet as pq
import pyarrow.dataset as ds
import sys

path="part-0.parquet"

if sys.argv[1] == "Dataset":
    print("Using Dataset.to_batches API")    
    batches = ds.dataset(path).to_batches(batch_size=100,
                                          batch_readahead=0,
                                          fragment_readahead=0)
else:
    print("Using ParquetFile.iter_batches API")
    batches = pq.ParquetFile(path).iter_batches(batch_size=100)

# Iterate through batches
max_alloc = 0
for batch in batches:
    alloc = pa.total_allocated_bytes()
    if alloc > max_alloc:
        max_alloc = alloc
        print("New max total_allocated_bytes", max_alloc)
del batches
print("Final:", pa.total_allocated_bytes())
```

I see the following numbers (pyarrow 16.1.0, python 3.11.6):

* Dataset.to_batches:
```shell
./pyarrow_39808_repro.py Dataset
Using Dataset.to_batches API
New max total_allocated_bytes 549322688
New max total_allocated_bytes 794176064
New max total_allocated_bytes 1094879616
New max total_allocated_bytes 1340710656
New max total_allocated_bytes 3732962048
New max total_allocated_bytes 4236869504
New max total_allocated_bytes 4849658496
New max total_allocated_bytes 6188106432
Final: 185074688
```

* ParquetFile.iter_batches:
```
./pyarrow_39808_repro.py ParquetFile
Using ParquetFile.iter_batches API     
New max total_allocated_bytes 252396608
New max total_allocated_bytes 252403456
<cut>
New max total_allocated_bytes 274519744
New max total_allocated_bytes 274521024
Final: 35072
```

i.e. `ParquetFile.iter_batches` requires at most ~261MB, whereas `Dataset.to_batches` requires ~5901MB - 22x more (!) - plus an additinal 176MB of RAM still in use after iteration complete.

**Comment by PatrikBernhard** [comment]
Is it possible to ask for an update on this issue? I am currently running into the same memory leak in my project which gets mitigated when using `Parquetfile.iter_batches` instead of `Dataset.to_batches`. However, I'd like to use the HivePartition-discovering within Dataset.

**Comment by mats-werrebrouck** [comment]
Is there any update on this issue? I am also running into the same memory leak and would like to stick with `Dataset.to_batches` because of the HivePartition-discovering within  PyArrow Dataset.

**Comment by bveeramani** [comment]
Hey, is there any update on this issue? It impacts Ray's `read_parquet` implementation: https://github.com/ray-project/ray/issues/49158

**Comment by alexeykudinkin** [comment]
@jorisvandenbossche would it be possible to clarify at least if the exhibited behavior is expected side-effect or not?

**Comment by HanxunH** [comment]
Any workaround? I'm also getting similar issues.

**Comment by wingkitlee0** [comment]
Came across this issue recently and I can still see this https://github.com/apache/arrow/issues/39808#issuecomment-2163183635

Previously I tried `pre_buffer=False` and `use_stream_buffer=True` in [`ParquetFragmentScanOptions` ](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.ParquetFragmentScanOptions.html), in which the `total_allocated_bytes` stopped growing.

There is also a new option `cache_metadata` in `to_batches` (not released yet; only dev version), which seems to reduce some %.

However, the "memory usage" difference between dataset and ParquetFile is still quite big.

**Comment by bwi-earth** [comment]
I'm having the very same problem (pyarrow 19.0.*)

i've tried:

- `ds.scanner(...).to_reader()`
- `ds.scanner(...).scan_batches()`
-  `ds.to_batches`

Memory is consistently growing, each time, at least memory shown by `memory_profiler` or `top`, the one  shown by `pa.pa.total_allocated_bytes()` is oscilating.
It seems its not related to any setting 

The source dataset is stored on s3

**Comment by jpfeuffer** [comment]
Hi! I am also kind of blocked by this issue. This is nr. 4 of the most voted issues (disregarding typing) and I was wondering if there is anything blocking from some of the core developers at least investigating this (apart from finding time obviously)? Can we maybe help by providing more examples?
Or are we missing some kind of workaround that makes this issue lower priority?

Or is it maybe not even a bug but an expected implementation detail as mentioned [here](https://github.com/ray-project/ray/issues/49158#issuecomment-2530009412)?

**Comment by akoumjian** [comment]
For what it's worth, my team still avoids using Dataset entirely due to this issue. Instead we have implemented our own file discovery, scanning, and concatenation. It is highly unlikely this is intended behavior, as this is not about peak memory usage as in that comment but unreleased memory many times greater than the uncompressed data should represent.

**Comment by bwi-earth** [comment]
The `ParquetFile.iter_batches` is not really a replacement since it does not allow scanning the row groups and retrieving only the relevant ones.

**Comment by wingkitlee0** [comment]
Sharing a plot that I made a little while ago, using pyarrow 20.x and the following options in `to_batches` (or `scanner`):

```
batch_readahead=0,
cache_metadata=False,  # new
fragment_scan_options=pyarrow.dataset.ParquetFragmentScanOptions(
  use_buffered_stream=True,
  pre_buffer=False,
  cache_options=pa.CacheOptions(lazy=True, prefetch_limit=0),
),
```

![Image](https://github.com/user-attachments/assets/9ee27d20-6c11-4c37-896e-1b0d3ce2b07c)

I used the 2.3gb parquet file earlier in the thread, which has about 10 row groups. In the figure, the blue lines are default option, orange/green are using `cache_metadata=False` etc. `b5000` etc are the batch size. The top panel shows the memory used by the current batch. The middle and bottom panels are pa.total_allocated_bytes and rss (from psutil), respectively.

There are 9-10 spikes, which seem to be at the beginning of each row group. The memory usage is still rising with the extra options, though much slower.

**Comment by jchia** [comment]
What's the status of this bug? It's been more than a year. The memory leak is an absolute show-stopper for using `pyarrow.dataset.dataset()` instead of the supposedly legacy `pyarrow.parquet.ParquetDataset()` on a dataset of substantial size. I'm sticking with pyarrow.parquet.ParquetDataset() for now.

**Comment by justinli500** [comment]
I've been working on a fix at the C++ Parquet reader level - evicting pre-buffered row-group bytes after they're consumed, so pre_buffer=True doesn't retain the full file in memory during iteration. It's a different approach from the recent scanner/Ray-side work (ray-project/ray#62745, b234d4e).

Opening a PR shortly - happy to coordinate if there's overlap I'm missing.
